index.bs

```diff
@@ -1046,84 +1005,431 @@ BigIntValue = {
+
+      1. Return [=success=] with data
+         <code>[=Object.fromEntries=](|deserialized key-value list|)</code>.
+
+    <dt>|type| equals <code>RegExpLocalValue.type</code>
+    <dd>Return [=success=] with data <code>new [=RegExp=](|value|)</code>.
+
+    <dt>|type| equals <code>SetLocalValue.type</code>
+    <dd>
+      1. Let |deserialized value list| be a result of [=try=]
+         [=map serialized value list to deserialized value list=] with given |value|.
+
+      1. If |deserialized value| is [=error=], return |deserialized value|.
+
+      1. Return [=success=] with data <code>new [=Set=](|deserialized values|)</code>.
+  </dl>
+
+1. Return [=error=] with [=error code=] [=invalid argument=]
+
+</div>
+
+### Remote Value ### {#data-types-protocolValue-RemoteValue}
+
+Values accessible from the ECMAScript runtime are represented by a mirror
+object, specified as <code>RemoteValue</code>. The value's type is specified in
+the <code>type</code> property. In the case of JSON-representable primitive
+values, this contains the value in the <code>value</code> property; in the case
+of non-JSON-representable primitives, the <code>value</code> property contains a
+string representation of the value. For non-primitive objects, the
+<code>objectId</code> property contains a string id that provides a unique
+handle to the object, valid for its lifetime inside the engine. For some
+non-primitive types, the <code>value</code> property contains a representation
+of the data in the ECMAScript object; for container types this can contain
+further <code>RemoteValue</code> instances. The <code>value</code> property can
+be null if there is a duplicate object i.e. the object has already been
+serialized in the current <code>RemoteValue</code>, perhaps as part of a
+cycle, or otherwise when the maximum serialization depth is reached.
+
+[=Nodes=] are also represented by <code>RemoteValue</code> instances. These have
+a partial serialization of the node in the value property.
+
+Note: mirror objects do not keep the original object alive in the runtime. If an
+object is discarded in the runtime, subsequent attempts to access it via the
+protocol will result in an error.
+
+<div algorithm>
+To get the <dfn>object id for an object</dfn> given a |session| and |object|:
+
+1. If |session|'s [=object id map=] does not contain |object|, run the
+   following steps:
+
+  1. Let |object id| be a new, unique, string identifier for |object|. If
+     |object| is an [=/element=] this must be the [=web element reference=] for
+     |object|; if it's a {{WindowProxy}} object, this must be the [=window
+     handle=] for |object|.
+
+  1. Set the value of |object| in |session|'s [=object id map=] to |object id|.
+
+1. Return the result of getting the value for |object| in |session|'s [=object
+   id map=].
+
+</div>
+
+[=remote end definition=] and [=local end definition=]
+```
+RemoteValue = {
+  PrimitiveProtocolValue //
```

Comment:
Should we call this `PrimitiveRemoteValue`?

Reply:
I think the idea is it's not representing a value that's transmitted only by copy, never by reference. Since the same types are by-copy going from remote-to-local and local-to-remote they can have a single representation in the schema. On the other hand by-reference values have an asymmetric representation; going from remote to local you need to transmit the type, and the handle, whereas going local to remote you only need to transmit the handle. For types where we support sending both a reference and a copy of the data (e.g. arrays, objects, regexp) remote to local sends the type, the handle, and a copy of the data, whereas local to remote sends either a handle _or_ the data required to construct a new object on the remote end.